geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/SimpleFeatureRecordConverter.scala

```diff
@@ -0,0 +1,195 @@
+    override def toRecord(attribute: UUID): String = if (attribute == null) { null } else { attribute.toString }
+    override def toAttribute(record: String): UUID = if (record == null) { null } else { UUID.fromString(record) }
+  }
+
+  object GeometryToRecordField {
+    def apply(name: String, encoding: GeometryEncoding): AttributeFieldConverter[Geometry, _] = {
+      encoding match {
+        case GeometryEncoding.Wkt => new GeometryToWktRecordField(name)
+        case GeometryEncoding.Wkb => new GeometryToWkbRecordField(name)
+        case _ => throw new NotImplementedError(s"Geometry encoding $encoding")
+      }
+    }
+  }
+
+  class GeometryToWktRecordField(name: String) extends AttributeFieldConverter[Geometry, String] {
+    override val field: RecordField = new RecordField(name, RecordFieldType.STRING.getDataType)
+    override def toRecord(attribute: Geometry): String = if (attribute == null) { null } else { WKTUtils.write(attribute) }
+    override def toAttribute(record: String): Geometry = if (record == null) { null } else { WKTUtils.read(record) }
+  }
+
+  class GeometryToWkbRecordField(name: String) extends AttributeFieldConverter[Geometry, Array[Byte]] {
+    override val field: RecordField = new RecordField(name, bytesType)
+    override def toRecord(attribute: Geometry): Array[Byte] = if (attribute == null) { null } else { WKBUtils.write(attribute) }
+    override def toAttribute(record: Array[Byte]): Geometry = if (record == null) { null } else { WKBUtils.read(record) }
+  }
+
+  class ListToRecordField(name: String, subType: AttributeFieldConverter[AnyRef, AnyRef])
+      extends AttributeFieldConverter[java.util.List[AnyRef], Array[AnyRef]] {
+    override val field: RecordField = new RecordField(name, RecordFieldType.ARRAY.getArrayDataType(subType.field.getDataType))
+    override def toRecord(attribute: java.util.List[AnyRef]): Array[AnyRef] =
+      if (attribute == null) { null } else { attribute.asScala.map(subType.toRecord).toArray }
+    override def toAttribute(record: Array[AnyRef]): java.util.List[AnyRef] =
+      if (record == null) { null } else { java.util.Arrays.asList(record.map(subType.toAttribute): _*) }
+  }
+
+  class MapToRecordField(name: String, valueType: AttributeFieldConverter[AnyRef, AnyRef])
+      extends AttributeFieldConverter[java.util.Map[AnyRef, AnyRef], java.util.Map[String, AnyRef]] {
+    override val field: RecordField =
+      new RecordField(name, RecordFieldType.MAP.getMapDataType(valueType.field.getDataType))
+    override def toRecord(attribute: java.util.Map[AnyRef, AnyRef]): java.util.Map[String, AnyRef] = {
+      if (attribute == null) { null } else {
+        attribute.asScala.map { case (k, v) => k.toString -> valueType.toRecord(v) }.asJava
+      }
+    }
+    override def toAttribute(record: java.util.Map[String, AnyRef]): java.util.Map[AnyRef, AnyRef] = {
+      if (record == null) { null } else {
+        // TODO records do not support non-string map keys
```

Comment:
Uhmm... is this TODO about heading userData which may have GeoTools Hints as keys?

Can I impose on you to write up a ticket with a line or two of details:)?

Reply:
actually this PR doesn't do anything with user data. this is for Map-type attributes in the simple feature. (we should probably just deprecate map-type attributes since they're not well supported and we can use json types instead)